lib/internal/modules/esm/assets/load.js

```diff
@@ -0,0 +1,305 @@
+      };
+      cacheForGET.set(parsed.href, entry);
+      fulfill(entry);
+    });
+  });
+  cacheForGET.set(parsed.href, result);
+  return result;
+}
+
+/**
+ * Returns if an address has local status by if it is going to a local
+ * interface or is an address resolved by DNS to be a local interface
+ * @param {string} hostname url.hostname to test
+ * @returns {Promise<boolean>}
+ */
+async function isLocalAddress(hostname) {
+  try {
+    if (StringPrototypeStartsWith(hostname, '[') &&
+        StringPrototypeEndsWith(hostname, ']')) {
+      hostname = StringPrototypeSlice(hostname, 1, -1);
+    }
+    const addr = await dnsLookup(hostname, { verbatim: true });
+    if (addr.family === 4) {
+      return addr.address === '127.0.0.1';
+    } else if (addr.family === 6) {
+      return normalizeIPv6(addr.address) ===
+        '0000:0000:0000:0000:0000:0000:0000:0001';
+    }
+  } catch {}
+  return false;
+}
+
+/**
+ * Used to expand interface addresses to a canonical form for comparisons
+ * E.G. comparing ::1, 0::1, etc. to 0000:0000:0000:0000:0000:0000:0000:0001
+ * Due to IPv6 having multiple ways to represent a single address we need to
+ * normalize
+ * @param {string} str
+ * @returns {string}
+ */
+function normalizeIPv6(str) {
+  const parts = StringPrototypeSplit(str, ':');
+  let expansionIndex = -1;
+  for (let i = 0; i < parts.length; i++) {
+    const part = parts[i];
+    if (part === '' && expansionIndex === -1 && parts.length < 8) {
+      expansionIndex = i;
+      continue;
+    }
+    parts[i] = StringPrototypePadStart(part, 4, '0');
+  }
+  if (parts.length < 8) {
+    ArrayPrototypeSplice(
+      parts,
+      expansionIndex,
+      1,
+      ...ArrayFrom({ length: 8 - parts.length + 1 }, () => '0000')
+    );
+  }
+  return ArrayPrototypeJoin(parts, ':');
+}
+
+/**
+ * @typedef CacheEntry
+ * @property resolvedHREF {Promise<string> | string}
+ * @property headers {Record<string, string>}
+ * @property body {Promise<Buffer> | Buffer}
+ */
+
+/**
+ * Only for GET requests, other requests would need new Map
+ * HTTP cache semantics keep diff caches
+ *
+ * Maps HREF to pending cache entry
+ * @type {Map<string, Promise<CacheEntry> | CacheEntry>}
+ */
+let cacheForGET = new SafeMap();
+
+/**
+ * Fetches a location with a shared cache following redirects.
+ * Does not respect HTTP cache headers.
+ *
+ * This splits the header and body Promises so that things only needing
+ * headers don't need to wait on the body.
+ * @param {URL} parsed
+ * @param {ESModuleContext} context
+ * @returns {ReturnType<typeof fetchWithRedirects>}
+ */
+function fetch(parsed, { parentURL }) {
+  const { href } = parsed;
+  const existing = cacheForGET.get(href);
+  if (existing) {
+    return existing;
+  }
+  if (parsed.protocol === 'http:') {
+    return isLocalAddress(parsed.hostname).then((is) => {
+      if (is !== true) {
+        let parent = fileURLToPath(parentURL);
+        if (parent.endsWith('[eval]')) parent = 'command-line';
```

Comment:
Anybody know why REPL doesn't support ESM? There's code in REPL specifically to support it, but specifying `--input-type=module` causes an error.

Reply:
seems out of scope of this PR, but I don't think REPL actually supports the Module parse goal anywhere?

Reply:
Adding/fixing REPL support of modules surely would be out of scope of this PR, but if it does work (or will before this PR lands), that'll need to be accounted for here (unless REPL also results in a filename of `'[eval]'` and thus this already covers it).